File: packages/typescript/x402/src/server/facilitator.ts
import { FacilitatorConfig, HTTPFacilitatorClient } from '@x402/core/server'
import { Network, PaymentRequirements } from '@x402/core/types'

const DEFAULT_FACILITATOR_URL = 'https://x402.4mica.xyz'

export interface OpenTabRequest {
  userAddress: string
  recipientAddress: string
  network?: Network
  erc20Token?: string
  ttlSeconds?: number
}

export interface OpenTabResponse {
  tabId: string
  userAddress: string
  recipientAddress: string
  assetAddress: string
  startTimestamp: number
  ttlSeconds: number
  nextReqId: string
}

export class OpenTabError extends Error {
  constructor(
    public readonly status: number,
    public readonly response: OpenTabResponse
  ) {
    super(`OpenTab failed with status ${status}`)
    this.name = 'OpenTabError'
  }
}

export class FourMicaFacilitatorClient extends HTTPFacilitatorClient {
  constructor(config?: FacilitatorConfig) {
    super({ ...config, url: config?.url ?? DEFAULT_FACILITATOR_URL })
  }

  async openTab(
    userAddress: string,
    paymentRequirements: PaymentRequirements,
    ttlSeconds?: number
  ): Promise<OpenTabResponse> {
    let headers: Record<string, string> = {
      'Content-Type': 'application/json',
    }

    const authHeaders = await this.createAuthHeaders('tabs')
    headers = { ...headers, ...authHeaders.headers }

    const response = await fetch(`${this.url}/tabs`, {
      method: 'POST',
      headers,
      body: JSON.stringify(
        this.safeJson({
          userAddress,
          recipientAddress: paymentRequirements.payTo,
          network: paymentRequirements.network,
          erc20Token: paymentRequirements.asset,
          ttlSeconds,
        })
      ),
    })

    const data = await response.json()

    if (typeof data === 'object' && data !== null && 'tabId' in data) {
      const openTabResponse = data as OpenTabResponse
      if (!response.ok) {
        throw new OpenTabError(response.status, openTabResponse)
      }
      return openTabResponse
    }

    throw new Error(`Facilitator openTab failed (${response.status}): ${JSON.stringify(data)}`)
  }

  /**
   * Helper to convert objects to JSON-safe format.
   * Handles BigInt and other non-JSON types.
   *
   * @param obj - The object to convert
   * @returns The JSON-safe representation of the object
   */
  private safeJson<T>(obj: T): T {
    return JSON.parse(
      JSON.stringify(obj, (_, value) => (typeof value === 'bigint' ? value.toString() : value))
    )
  }
}

File: packages/rust/x402/src/server/facilitator.rs
use std::sync::Arc;

use http::StatusCode;
use reqwest::Client;
use serde::{Deserialize, Serialize};
use x402_axum::facilitator_client::{FacilitatorClient, FacilitatorClientError};

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct OpenTabRequest {
    pub user_address: String,
    pub recipient_address: String,
    pub network: Option<String>,
    pub erc20_token: Option<String>,
    pub ttl_seconds: Option<u64>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct OpenTabResponse {
    pub tab_id: String,
    pub user_address: String,
    pub recipient_address: String,
    pub asset_address: String,
    pub start_timestamp: i64,
    pub ttl_seconds: i64,
    pub next_req_id: Option<String>,
}

pub struct TabFacilitatorClient {
    base: Arc<FacilitatorClient>,
    client: Client,
}

impl TabFacilitatorClient {
    pub fn new(base: Arc<FacilitatorClient>) -> Self {
        Self {
            base,
            client: Client::new(),
        }
    }

    /// Sends a `POST /tabs` request to the facilitator.
    pub async fn open_tab(
        &self,
        request: &OpenTabRequest,
    ) -> Result<OpenTabResponse, FacilitatorClientError> {
        let tabs_url =
            self.base
                .base_url()
                .join("./tabs")
                .map_err(|e| FacilitatorClientError::UrlParse {
                    context: "Failed to construct ./tabs URL",
                    source: e,
                })?;

        let mut req = self.client.post(tabs_url).json(request);
        for (key, value) in self.base.headers().iter() {
            req = req.header(key, value);
        }
        if let Some(timeout) = *self.base.timeout() {
            req = req.timeout(timeout);
        }

        let http_response = req.send().await.map_err(|e| FacilitatorClientError::Http {
            context: "POST /tabs",
            source: e,
        })?;

        if http_response.status() == StatusCode::OK {
            http_response.json::<OpenTabResponse>().await.map_err(|e| {
                FacilitatorClientError::JsonDeserialization {
                    context: "POST /tabs",
                    source: e,
                }
            })
        } else {
            let status = http_response.status();
            let body = http_response.text().await.map_err(|e| {
                FacilitatorClientError::ResponseBodyRead {
                    context: "POST /tabs",
                    source: e,
                }
            })?;
            Err(FacilitatorClientError::HttpStatus {
                context: "POST /tabs",
                status,
                body,
            })
        }
    }
}
